### PR TITLE
ci: e2e tests on windows

### PR DESCRIPTION
**Pull request by dpilch**
#### Description of changes

Run E2E tests on Windows.

The Windows bash terminal on CodeBuild appears to be non-interactive. This is causing issues with selecting a region when running `amplify configure`. I've set all the Windows E2E tests to run `us-east-1` for now and we can fix that later when you have some more time.

There are two disabled tests for windows only:
1. amplify pull on an ios project.
2. Cypress run after generating graphql statements.

Test 1 is not very valuable. It is possible to do this with the current CLI, but realistically no customer would do this because XCode is not available on windows.

Test 2 is much more valuable. This asserts that the generated graphql statements work in a running app. The app compiles but then there is an error in cypress. It does not appear to be an error in a test case. It might be an issue with cypress on the windows instance.

I was not able to reproduce either case on my own windows workspace. 

To recap, we have 3 follow ups to complete later:
1. Windows E2E tests on other regions
2. Enable amplify pull on ios project test
3. Enable Cypress run after generating graphql statements test.

#### Issue #, if available

N/A

#### Description of how you validated changes

Run the E2E tests.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-codegen/blob/main/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .codebuild/scripts/e2e_test_windows.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+# set exit on error to true
+set -e
+
+source ./shared-scripts.sh && _setupE2ETestsWindows
+codebuild-breakpoint
```
Comment (dpilch) [inline .codebuild/scripts/e2e_test_windows.sh:7]: This is keeping the same behavior as E2E tests on linux. I'm not sure `codebuild-breakpoint` is strictly needed on either.

File: packages/amplify-codegen-e2e-core/src/categories/api.ts [hunk 2]
```diff
@@ -49,9 +49,7 @@ export function addApiWithoutSchema(cwd: string, opts: Partial<AddApiOptions & {
       .sendCarriageReturn()
       .wait('Do you want to edit the schema now?')
       .sendConfirmNo()
-      .wait(
-        '"amplify publish" will build all your local backend and frontend resources (if you have hosting category added) and provision it in the cloud',
-      )
+      .wait('"amplify publish" will build all your local backend and frontend resources')
```
Comment (dpilch) [inline packages/amplify-codegen-e2e-core/src/categories/api.ts:52]: Strange error.

```
● codegen remove tests - iOS › Does not delete files during codegen remove
--
351 |  
352 | Expected value   "\"amplify publish\" will build all your local backend and frontend resources (if you have hosting category added) and provision it in the cloud"
353 | Received:
354 | "ision it in the cloud"
355 |  
356 | Message:
357 | expected "ision it in the cloud" to contain "\"amplify publish\" will build all your local backend and frontend resources (if you have hosting category added) and provision it in the cloud"
358 |  
359 | Difference:
360 |  
361 | - Expected
362 | + Received
363 |  
364 | - "amplify publish" will build all your local backend and frontend resources (if you have hosting category added) and provision it in the cloud
365 | + ision it in the cloud
```

The message is correct when running manually. Maybe there is a maximum length? The "v" in "provision" is the 120th character.

This change was also made on the CLI E2E when enabling windows.
https://github.com/aws-amplify/amplify-cli/commit/97c6de42f906d56a5fd0d78c4827110508955b93

File: shared-scripts.sh [hunk 3]
```diff
@@ -235,18 +244,32 @@ function _setupE2ETestsLinux {
...
     retry runE2eTest
 }
 
+function _runE2ETestsWindows {
+    echo "RUN E2E Tests Windows"
+    retry runE2eTest
+}
+
 function _scanArtifacts {
-    if ! yarn ts-node .codebuild/scripts/scan_artifacts.ts; then
```
Comment (dpilch) [inline shared-scripts.sh:249]: Get an error when using yarn over npx/npm. It looks like there are some workarounds, but it is easier to just use npx/npm here because there will be no functional difference to the scripts.

```
error Couldn't find the binary C:\codebuild\tmp\output\src2004\src\github.com\aws-amplify\amplify-codegen\node_modules\.bin\ts-node
```

File: .codebuild/scripts/local_publish_helpers.sh [hunk 4]
```diff
@@ -7,7 +7,7 @@ function startLocalRegistry {
     # Start local registry
     tmp_registry_log="$(mktemp)"
     echo "Registry output file: $tmp_registry_log"
-    (cd && nohup npx ${VERDACCIO_PACKAGE:-$default_verdaccio_package} -c $1 &>$tmp_registry_log &)
+    nohup npx ${VERDACCIO_PACKAGE:-$default_verdaccio_package} -c $1 &>$tmp_registry_log &
```
Comment (dpilch) [inline .codebuild/scripts/local_publish_helpers.sh:10]: The `cd` is unnecessary here and caused some issues on Windows.

File: packages/amplify-codegen-e2e-core/package.json [hunk 5]
```diff
@@ -32,7 +32,7 @@
     "jest-circus": "^27.5.1",
     "jest-environment-node": "^26.6.2",
     "lodash": "^4.17.19",
-    "node-pty": "beta",
+    "node-pty": "1.0.0",
```
Comment (dpilch) [inline packages/amplify-codegen-e2e-core/package.json:35]: This might not be necessary for this change. I tried this when troubleshooting various issues. However, this is probably something we should do regardless so that our version is stable. I can try removing this and adding it to a separate change if it is desired.

File: scripts/split-e2e-tests.ts [hunk 6]
```diff
@@ -71,7 +71,11 @@ type CandidateJob = {
   runSolo: boolean;
 };
 const createJob = (os: OS_TYPE, jobIdx: number, runSolo: boolean = false): CandidateJob => {
-  const region = AWS_REGIONS_TO_RUN_TESTS[jobIdx % AWS_REGIONS_TO_RUN_TESTS.length];
+  // The bash terminal for Windows on CodeBuild is non-interactive.
+  // amplify-configure will always choose us-east-1 for the region.
+  // This will set all Windows jobs to use us-east-1 to avoid region mismatch.
+  // We will come back to this later to properly fix the testing issue.
+  const region = os === 'w' ? 'us-east-1' : AWS_REGIONS_TO_RUN_TESTS[jobIdx % AWS_REGIONS_TO_RUN_TESTS.length];
```
Comment (dpilch) [inline scripts/split-e2e-tests.ts:78]: Another weird thing, this only seems to affect the `amplify configure` step. On command like `codegen add` the selection is set correctly.

**Review by alharris-at — APPROVED** [review]

File: .codebuild/scripts/e2e_test_windows.sh (hunk 1 above)
Comment (phani-srikar) [reply .codebuild/scripts/e2e_test_windows.sh:7]: This is used when you want to run the tests in debug mode. I think the script is called `yarn-e2e-debug`.

File: packages/amplify-codegen-e2e-core/src/categories/api.ts (hunk 2 above)
Comment (phani-srikar) [reply packages/amplify-codegen-e2e-core/src/categories/api.ts:52]: Alternatively, you can wrap the message with `*/` regex to ignore subsequent parts. Just FYI.

File: packages/amplify-codegen-e2e-tests/src/__tests__/graphql-documents-generator.test.ts [hunk 7]
```diff
@@ -31,9 +32,10 @@ describe('GraphQL documents generator e2e tests', () => {
   });
 
   const schemaFileName = 'schema.graphql';
-  it('generates valid GraphQL documents for given schema', async () => {
+  // skip cypress test on windows
+  (isWindows() ? it.skip : it)('generates valid GraphQL documents for given schema', async () => {
```
Comment (phani-srikar) [inline packages/amplify-codegen-e2e-tests/src/__tests__/graphql-documents-generator.test.ts:36]: Do we want to run these on windows as well eventually? If yes, we should have a task for it.

File: scripts/split-e2e-tests.ts (hunk 6 above)
Comment (phani-srikar) [reply scripts/split-e2e-tests.ts:78]: We might need to look into how we're sending the region selection command interactively and make it for both platforms. Running in `us-east-1` is fine for now.

**Review by phani-srikar — APPROVED** [review]

File: scripts/split-e2e-tests.ts (hunk 6 above)
Comment (dpilch) [reply scripts/split-e2e-tests.ts:78]: It is using j/k to navigate the list. I tried with up/down arrow as well and ran into the same issue. I've created a task to investigate further.

File: packages/amplify-codegen-e2e-tests/src/__tests__/graphql-documents-generator.test.ts (hunk 7 above)
Comment (dpilch) [reply packages/amplify-codegen-e2e-tests/src/__tests__/graphql-documents-generator.test.ts:36]: I've created tasks for the 2 disabled tests and the region issue.
